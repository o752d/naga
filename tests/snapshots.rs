use serde::Deserialize;

#[derive(Hash, PartialEq, Eq, Deserialize)]
enum Stage {
    Vertex,
    Fragment,
    Compute,
}

#[derive(Hash, PartialEq, Eq, Deserialize)]
struct BindSource {
    stage: Stage,
    group: u32,
    binding: u32,
}

#[derive(Deserialize)]
struct BindTarget {
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    #[serde(default)]
    buffer: Option<u8>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    #[serde(default)]
    texture: Option<u8>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    #[serde(default)]
    sampler: Option<u8>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    #[serde(default)]
    mutable: bool,
}

#[derive(Default, Deserialize)]
struct Parameters {
    #[serde(default)]
    #[allow(dead_code)]
    spv_flow_dump_prefix: String,
    #[cfg_attr(not(feature = "spv-out"), allow(dead_code))]
    spv_capabilities: naga::FastHashSet<spirv::Capability>,
    #[cfg_attr(not(feature = "msl-out"), allow(dead_code))]
    mtl_bindings: naga::FastHashMap<BindSource, BindTarget>,
}

bitflags::bitflags! {
    struct Language: u32 {
        const SPIRV = 0x1;
        const METAL = 0x2;
        const GLSL = 0x4;
    }
}

#[cfg(feature = "spv-out")]
fn check_output_spv(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::spv;
    use rspirv::binary::Disassemble;

    let spv = spv::write_vec(
        &module,
        spv::WriterFlags::NONE,
        params.spv_capabilities.clone(),
    )
    .unwrap();

    let dis = rspirv::dr::load_words(spv)
        .expect("Produced invalid SPIR-V")
        .disassemble();
    insta::assert_snapshot!(format!("{}.spvasm", name), dis);
}

#[cfg(feature = "msl-out")]
fn check_output_msl(module: &naga::Module, name: &str, params: &Parameters) {
    use naga::back::msl;

    let mut binding_map = msl::BindingMap::default();
    for (key, value) in params.mtl_bindings.iter() {
        binding_map.insert(
            msl::BindSource {
                stage: match key.stage {
                    Stage::Vertex => naga::ShaderStage::Vertex,
                    Stage::Fragment => naga::ShaderStage::Fragment,
                    Stage::Compute => naga::ShaderStage::Compute,
                },
                group: key.group,
                binding: key.binding,
            },
            msl::BindTarget {
                buffer: value.buffer,
                texture: value.texture,
                sampler: value.sampler,
                mutable: value.mutable,
            },
        );
    }
    let options = msl::Options {
        lang_version: (1, 0),
        spirv_cross_compatibility: false,
        binding_map,
    };

    let (msl, _) = msl::write_string(&module, &options).unwrap();
    insta::assert_snapshot!(format!("{}.msl", name), msl);
}

#[cfg(feature = "glsl-out")]
fn check_output_glsl(module: &naga::Module, name: &str, stage: naga::ShaderStage, ep_name: &str) {
    use naga::back::glsl;

    let options = glsl::Options {
        version: glsl::Version::Embedded(310),
        entry_point: (stage, ep_name.to_string()),
    };

    let mut buffer = Vec::new();
    let mut writer = glsl::Writer::new(&mut buffer, &module, &options).unwrap();
    writer.write().unwrap();

    let string = String::from_utf8(buffer).unwrap();
    insta::assert_snapshot!(format!("{}-{:?}.glsl", name, stage), string);
}

#[cfg(feature = "wgsl-in")]
fn convert_wgsl(name: &str, language: Language) {
    let params =
        match std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".param.ron")) {
            Ok(string) => ron::de::from_str(&string).expect("Couldn't find param file"),
            Err(_) => Parameters::default(),
        };

    let module = naga::front::wgsl::parse_str(
        &std::fs::read_to_string(format!("tests/snapshots/in/{}{}", name, ".wgsl"))
            .expect("Couldn't find wgsl file"),
    )
    .unwrap();
    naga::proc::Validator::new().validate(&module).unwrap();

    #[cfg(feature = "spv-out")]
    {
        if language.contains(Language::SPIRV) {
            check_output_spv(&module, name, &params);
        }
    }
    #[cfg(feature = "msl-out")]
    {
        if language.contains(Language::METAL) {
            check_output_msl(&module, name, &params);
        }
    }
    #[cfg(feature = "glsl-out")]
    {
        if language.contains(Language::GLSL) {
            for &(stage, ref ep_name) in module.entry_points.keys() {
                check_output_glsl(&module, name, stage, ep_name);
            }
        }
    }
}

#[cfg(feature = "wgsl-in")]
#[test]
fn converts_wgsl_quad() {
    convert_wgsl("quad", Language::all());
}

#[cfg(feature = "wgsl-in")]
#[test]
fn converts_wgsl_simple() {
    convert_wgsl("simple", Language::all());
}

#[cfg(feature = "wgsl-in")]
#[test]
fn converts_wgsl_boids() {
    convert_wgsl("boids", Language::METAL);
}
