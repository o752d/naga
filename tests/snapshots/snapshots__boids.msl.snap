---
source: tests/snapshots.rs
expression: msl
---
#include <metal_stdlib>
#include <simd/simd.h>

typedef float2 type;
typedef float4 type1;
typedef float type2;
struct Particle {
	type pos;
	type vel;
};
struct SimParams {
	type2 deltaT;
	type2 rule1Distance;
	type2 rule2Distance;
	type2 rule3Distance;
	type2 rule1Scale;
	type2 rule2Scale;
	type2 rule3Scale;
};
typedef uint type3;
typedef Particle type4[5];
struct Particles {
	type4 particles;
};
typedef uint3 type5;
typedef int type6;

struct main1Input {
	type a_particlePos [[attribute(0)]];
	type a_particleVel [[attribute(1)]];
	type a_pos [[attribute(2)]];
};
struct main1Output {
	type1 gl_Position [[position]];
};
vertex main1Output main1(
	main1Input input [[stage_in]]
) {
	main1Output output;
	output.gl_Position = float4(float2(input.a_pos.x * metal::cos(-metal::atan2(input.a_particleVel.x, input.a_particleVel.y)) - input.a_pos.y * metal::sin(-metal::atan2(input.a_particleVel.x, input.a_particleVel.y)), input.a_pos.x * metal::sin(-metal::atan2(input.a_particleVel.x, input.a_particleVel.y)) + input.a_pos.y * metal::cos(-metal::atan2(input.a_particleVel.x, input.a_particleVel.y))) + input.a_particlePos, 0.0, 1.0);
	return output;
}
struct main2Input {
};
struct main2Output {
	type1 fragColor [[color(0)]];
};
fragment main2Output main2(
	main2Input input [[stage_in]]
) {
	main2Output output;
	output.fragColor = float4(1.0, 1.0, 1.0, 1.0);
	return output;
}
kernel void main3(
	constant SimParams& params [[buffer(0)]],
	constant Particles& particlesA [[buffer(1)]],
	device Particles& particlesB [[buffer(2)]],
	type5 gl_GlobalInvocationID [[thread_position_in_grid]]
) {
	type vPos;
	type vVel;
	type cMass;
	type cVel;
	type colVel;
	type6 cMassCount = 0;
	type6 cVelCount = 0;
	type pos1;
	type vel1;
	type3 i = 0;
	if (gl_GlobalInvocationID.x >= static_cast<uint>(5)) {
	}
	vPos = particlesA.particles[gl_GlobalInvocationID.x].pos;
	vVel = particlesA.particles[gl_GlobalInvocationID.x].vel;
	cMass = float2(0.0, 0.0);
	cVel = float2(0.0, 0.0);
	colVel = float2(0.0, 0.0);
	while(true) {
		if (i >= static_cast<uint>(5)) {
			break;
		}
		if (i == gl_GlobalInvocationID.x) {
			continue;
		}
		pos1 = float2(particlesA.particles[i].pos.x, particlesA.particles[i].pos.y);
		vel1 = float2(particlesA.particles[i].vel.x, particlesA.particles[i].vel.y);
		if (metal::distance(pos1, vPos) < params.rule1Distance) {
			cMass = cMass + pos1;
			cMassCount = cMassCount + 1;
		}
		if (metal::distance(pos1, vPos) < params.rule2Distance) {
			colVel = colVel - pos1 - vPos;
		}
		if (metal::distance(pos1, vPos) < params.rule3Distance) {
			cVel = cVel + vel1;
			cVelCount = cVelCount + 1;
		}
	}
	if (cMassCount == 0) {
		cMass = cMass / float2(cMassCount, cMassCount) + vPos;
	}
	if (cVelCount == 0) {
		cVel = cVel / float2(cVelCount, cVelCount);
	}
	vVel = vVel + cMass * params.rule1Scale + colVel * params.rule2Scale + cVel * params.rule3Scale;
	vVel = metal::normalize(vVel) * metal::clamp(metal::length(vVel), 0.0, 0.1);
	vPos = vPos + vVel * params.deltaT;
	if (vPos.x < -1.0) {
		vPos.x = 1.0;
	}
	if (vPos.x == 1.0) {
		vPos.x = -1.0;
	}
	if (vPos.y < -1.0) {
		vPos.y = 1.0;
	}
	if (vPos.y == 1.0) {
		vPos.y = -1.0;
	}
	particlesB.particles[gl_GlobalInvocationID.x].pos = vPos;
	particlesB.particles[gl_GlobalInvocationID.x].vel = vVel;
}

